File: ml/src/wikidetect/stance.py
"""Python view of shared/stance.json — the same vocabulary the Node eval
harness compiles (web/lib/stance.js). Data is shared; the ~15 lines of
compile logic are duplicated per language by design (no cross-language
imports)."""

import json
import re

from . import config

_QUOTE_RE = re.compile(r"[“\"][^“”\"]*[”\"]")


def _compile():
    raw = json.loads((config.REPO / "shared" / "stance.json").read_text())
    cats = []
    for cat in raw["categories"]:
        if "terms" in cat:
            terms = sorted(cat["terms"], key=len, reverse=True)
            pattern = r"\b(?:" + "|".join(re.escape(t) for t in terms) + r")\b"
            cats.append((cat["key"], re.compile(pattern, re.I)))
        else:
            flags = re.I if "i" in cat.get("flags", "") else 0
            cats.append((cat["key"], re.compile(cat["regex"], flags)))
    return cats


_CATEGORIES = None


def without_quotes(s: str) -> str:
    """Kept quotations may contain evaluative language; blank them out so
    only the article's own voice is checked."""
    return _QUOTE_RE.sub("“…”", s)


def stance_hits(text: str) -> list[dict]:
    global _CATEGORIES
    if _CATEGORIES is None:
        _CATEGORIES = _compile()
    prose = without_quotes(text)
    found = []
    for key, pattern in _CATEGORIES:
        hits = sorted(set(pattern.findall(prose)))
        if hits:
            found.append({"key": key, "hits": hits})
    return found

File: web/lib/stance.js
// Compiles shared/stance.json into the STANCE categories the eval harness
// and exemplar harvester assert with. Node-only (reads from disk) — the
// browser fetches shared/stance.json itself (app.js) for display-only
// highlighting; the vocabulary is never prompt material (see rules.js for why).
//
// To tighten what the evals catch, edit shared/stance.json — that file is
// the single source of the vocabulary for both the Node and Python sides.

import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';

const raw = JSON.parse(
  readFileSync(fileURLToPath(new URL('../../shared/stance.json', import.meta.url)), 'utf8')
);

const escRe = s => s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
const termRegex = (terms, flags = 'gi') =>
  new RegExp('\\b(?:' + [...terms].sort((a, b) => b.length - a.length).map(escRe).join('|') + ')\\b', flags);

export const STANCE = raw.categories.map(cat =>
  cat.terms
    ? { key: cat.key, terms: cat.terms, re: termRegex(cat.terms) }
    : { key: cat.key, re: new RegExp(cat.regex, cat.flags) }
);

// Kept quotations are allowed to contain evaluative language; blank them out
// before checking for stance so only the article's own voice is asserted.
export const withoutQuotes = s => s.replace(/[“"][^“”"]*[”"]/g, '“…”');

// All stance matches in `text` (quotes blanked), as [{key, hits}].
export function stanceHits(text) {
  const prose = withoutQuotes(text);
  const found = [];
  for (const cat of STANCE) {
    const m = prose.match(new RegExp(cat.re.source, cat.re.flags));
    if (m) found.push({ key: cat.key, hits: [...new Set(m)] });
  }
  return found;
}
